### PR TITLE
Audio test

### PR DESCRIPTION
**Pull request by mikebeaton**
@vit9696 , @Goldfish64 - if you had a moment to review some bits of this, I made some more proposed changes to Audio.

It is split into separate small commits to make it easier to review, I hope.

The only two changes I'd really appreciate some feedback on are:

 - @vit9696, @Goldfish64 : be54db75394ad39c6d3863c1d6bb35d8dc00f1d1 , I have not just moved the repeated 'reset pci attributes' code into the shared call which it previously came before in two places, but also moved that code to the end of the shared method, even though it was called before it, i.e. I am now resetting to the original PCI state at a _different_ time to when it was done before. According to this commit, any previous PCI settings are now restored after the rest of the controller clear down, since they are set before the rest of the controller set up; I wanted to double-check that I haven't missed any reason why this is the wrong ordering
 - @vit9696 : b375de3dbf38f0c54023107c317e42e3075c333c , I've introduced 3 PCDs for AudioDxe
   -  Firstly, perhaps you could double-check that I've introduced the PCDs correctly (though I think I have), but mainly:
   - `PcdAudioCodecErrorOnNoOutputs`: pretty sure this one makes sense, and that the code this introduces is neater than the `#define` I had used before
   - `PcdAudioControllerTryProtocolGetMode`: I think this one is good, as it preserves your code to try EFI_OPEN_PROTOCOL_GET_PROTOCOL as a fallback after EFI_OPEN_PROTOCOL_BY_DRIVER, even though it not turned on (as I don't _think_ we need it, with 'DisconnectHda' quirk)
   - `PcdAudioControllerResetOnExitBootServices`: least sure about this one, there was an (unused) var in the controller struct to hold an Exit Boot Services event already (@Goldfish64, perhaps if you have time, you might want to check that I've used it for more or less what you had planned for it?), but even though I have now implemented this - to try resetting the controller on exit, to try to resolve the no-sound-after-AudioDxe-in-Windows-on-a-few-machines issue (and @Goldfish64 suggested the same idea on OC private discussion thread, if I understood him correctly, I think I did) - this did not actually make any difference to that problem, unfortunately; so not sure whether to leave this more or less untested/not-proven-worthwhile code in there on a turned-off PCD or not?

### CODE REVIEW

**Comment by mikebeaton** [comment]
@vit9696 : there is actually a not-quite-related commit in there ef34d463a5c19e359794821303fb477ea2276c46 to add some more logging (either VERBOSE or error-condition only) in OcExitBootServices, particularly to log a warning if it won't exit.

As you're no doubt well aware, I found if you do something illegal in OcExitBootServices (e.g. I was trying to disconnect the sound controller in AudioExitBootServices....) and thereore OcExitBootServices, and so ExitBootServices, fails, then all OSes (macOS, Linux, Windows) retry gBS->ExitBootServices (and therefore re-call OcExitBootServices) (a bit like ForceExitBootServices, but it is not ForceExitBootServices) at least once, sometimes more when I added some test code to fail on purpose; which means it's possible for gBS->ExitBootServices to fail silently once then succeed, so I thought adding a warning to the log when it fails was probably useful.

File: Library/OcAfterBootCompatLib/ServiceOverrides.c [hunk 1]
```diff
@@ -964,10 +966,14 @@ OcExitBootServices (
   // For non-macOS operating systems return directly.
   //
   if (BootCompat->ServiceState.AppleBootNestedCount == 0) {
-    return BootCompat->ServicePtrs.ExitBootServices (
+    Status = BootCompat->ServicePtrs.ExitBootServices (
       ImageHandle,
       MapKey
       );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_WARN, "OCABC: OcExitBootServices (%amacOS) - %r\n", "non-", Status));
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: You are not allowed to print within/after ExitBootServices. Even on failure.

File: Library/OcAfterBootCompatLib/ServiceOverrides.c [hunk 2]
```diff
@@ -995,6 +1001,7 @@ OcExitBootServices (
   // Abort on error.
   //
   if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_WARN, "OCABC: OcExitBootServices (%amacOS) - %r\n", "", Status));
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/ServiceOverrides.c:1004]: Same as above.

File: Library/OcAfterBootCompatLib/ServiceOverrides.c (hunk 1 above)
Comment (mikebeaton) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: https://github.com/acidanthera/OpenCorePkg/blob/master/Library/OcMainLib/OpenCoreUefiInOut.c#L63-L67

Comment (mikebeaton) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: Actually, I guess the above is not (strictly) within or after, so is different?

Comment (mikebeaton) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: Actually, again, no - the above is still after the memory map has been taken - so not entirely clear of the difference... ... I suppose in this case here, the memory map has been taken, and the debug lib might have been torn down? (Probably the other comment is sightly misleading still, though? Or could do with a new comment - maybe?? - where I tried to add some code?)

File: Library/OcAfterBootCompatLib/ServiceOverrides.c (hunk 2 above)
Comment (mikebeaton) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:1004]: Yes, agreed, as above.

File: Library/OcAfterBootCompatLib/ServiceOverrides.c (hunk 1 above)
Comment (mikebeaton) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: I won't push this - ofc! - and I'll offer sightly updated comments in as separate commit and see what you think.

**Comment by Sher1ocks** [comment]
hi mike. i report log message.

my desktop board is asus b150m.

File: Library/OcAfterBootCompatLib/ServiceOverrides.c (hunk 1 above)
Comment (vit9696) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:974]: The example is different, as the print is almost impossible to appear, and is a bug. Here the print will frequently trigger on many machines and will bork them.

File: Library/OcDriverConnectionLib/DriverDisconnection.c [hunk 3]
```diff
@@ -243,8 +246,17 @@ OcDisconnectHdaControllers (
         continue;
       }
 
+      DevicePath = ConvertDevicePathToText (
```
Comment (vit9696) [inline Library/OcDriverConnectionLib/DriverDisconnection.c:249]: Use DebugPrintPath in my opinion.

**Comment by mikebeaton** [comment]
> hi @mikebeaton . i report log message.  my desktop mainboard is asus b150m.

@Sher1ocks Thank you for the report. Just to be clear, which versions does this occur on?
 - [ ] this `audio-test` branch?
 - [ ] current `master` branch with work-in-progress 0.7.7 after recent changes to AudioDxe?
 - [ ] 0.7.6 release before any changes to AudioDxe?

**Comment by Sher1ocks** [comment]
> > hi @mikebeaton . i report log message.  my desktop mainboard is asus b150m.
> 
> @Sher1ocks Thank you for the report. Just to be clear, which versions does this occur on?
> 
> * [ ]  this `audio-test` branch?
> * [ ]  current `master` branch with work-in-progress 0.7.7 after recent changes to AudioDxe?
> * [ ]  0.7.6 release before any changes to AudioDxe?

1. no audio-test branch result. just master branch

**Comment by Sher1ocks** [comment]
sorry my mistake close button in mobile. please back reopen.

**Comment by mikebeaton** [comment]
> > > hi @mikebeaton . i report log message.  my desktop mainboard is asus b150m.
> > 
> > 
> > @Sher1ocks Thank you for the report. Just to be clear, which versions does this occur on?
> > 
> > * [ ]   this `audio-test` branch?
> > * [ ]   current `master` branch with work-in-progress 0.7.7 after recent changes to AudioDxe?
> > * [ ]   0.7.6 release before any changes to AudioDxe?
> 
> 1. no audio-test branch result. just lastest master branch
> 2. yes.
> 3. there is no log like attached pic in 0.7.6

Okay, there is no log like attached pic in 0.7.6, but again just to be clear: did the driver actually work for you before in 0.7.6, i.e. did it produce sound for you before (in 0.7.6) but now (0.7.7) it does not? Or is it just a difference in what is logged?

**Comment by Sher1ocks** [comment]
> > 1. no audio-test branch result. just lastest master branch
> > 2. yes.
> > 3. there is no log like attached pic in 0.7.6
> 
> Okay, there is no log like attached pic in 0.7.6, but again just to be clear: did the driver actually work for you before in 0.7.6, i.e. did it produce sound for you before (in 0.7.6) but now (0.7.7) it does not? Or is it just a difference in what is logged?

@mikebeaton 
1. 0.7.7(also 0.7.6) sound is working before this 706cb4e AudioDxe: Updates
2. 0.7.7(also 0.7.6) there is log, but since AudioDxe: Updates, i always get time out log.

**Comment by mikebeaton** [comment]
@Sher1ocks Thanks. Can you send me a debug boot log from before and after the commit which breaks sound for you?

**Comment by Sher1ocks** [comment]
> @Sher1ocks Thanks. Can you send me a debug boot log from before and after the commit which breaks sound for you?
can you use element? i'm in online now

File: Library/OcAfterBootCompatLib/ServiceOverrides.c [hunk 4]
```diff
@@ -927,6 +927,17 @@ OcStartImage (
...
+  Notes:
+    - Most OSes attempt to call ExitBootServices more than once if it fails initially
+      (similar to OpenCore ForceExitBootServices)
+      - Therefore, OcExitBootServices may get called more than once
+      - However this should never be relied upon for correct operation
+    - Any logging within this call but before original ExitBootServices is attempted
+      (e.g. within a scheduled handler) may cause ExitBootServices to fail (e.g. it may
+      change the memory map by allocating), and should only be done, if at all, in the
+      case of unexpected errors
+    - Never log after original ExitBootServices has been attempted, not even on error
```
Comment (mikebeaton) [inline Library/OcAfterBootCompatLib/ServiceOverrides.c:940]: @vit9696 Are these comments correct, and do you think worth having if so?

File: Library/OcDebugLogLib/DebugPrint.c [hunk 5]
```diff
@@ -36,6 +36,23 @@ DebugPrintDevicePath (
...
+  DEBUG_CODE_BEGIN ();
+  
+  EFI_DEVICE_PATH_PROTOCOL  *DevicePath;
+
+  DevicePath = DevicePathFromHandle (Handle);
+  DebugPrintDevicePath (ErrorLevel, Message, DevicePath);
+
+  DEBUG_CODE_END ();
+}
+
```
Comment (mikebeaton) [inline Library/OcDebugLogLib/DebugPrint.c:55]: Also just to double-check this is okay to add?

File: Library/OcAfterBootCompatLib/ServiceOverrides.c (hunk 4 above)
Comment (vit9696) [reply Library/OcAfterBootCompatLib/ServiceOverrides.c:940]: Yes.

File: Library/OcDebugLogLib/DebugPrint.c (hunk 5 above)
Comment (vit9696) [reply Library/OcDebugLogLib/DebugPrint.c:55]: DebugPrintDevicePath does not take OPTIONAL DevicePath, so better to check the result of DevicePathFromHandle for NULL. This change will need to be added to all Debug libraries in OpenCore.

File: Staging/AudioDxe/HdaController/HdaController.c [hunk 6]
```diff
@@ -324,10 +330,31 @@ HdaControllerGetName (
   DEBUG ((DEBUG_INFO, "HDA: Controller is %s\n", HdaControllerDev->Name));
 }
 
+STATIC
+VOID
+EFIAPI
+HdaControllerExitBootServicesHandler (
```
Comment (vit9696) [inline Staging/AudioDxe/HdaController/HdaController.c:336]: Need to clarify this is for Windows I believe?

File: Staging/AudioDxe/HdaController/HdaController.c [hunk 7]
```diff
@@ -364,6 +391,24 @@ HdaControllerReset (
     }
   }
 
+  if (!Restart) {
+    return EFI_SUCCESS;
+  }
+
+  if (PcdGetBool (PcdAudioControllerResetPciOnExitBootServices) &&
```
Comment (vit9696) [inline Staging/AudioDxe/HdaController/HdaController.c:398]: We usually put && to the next line.

File: Library/OcDebugLogLib/DebugPrint.c [hunk 8]
```diff
@@ -20,14 +20,17 @@ VOID
 DebugPrintDevicePath (
   IN UINTN                     ErrorLevel,
   IN CONST CHAR8               *Message,
-  IN EFI_DEVICE_PATH_PROTOCOL  *DevicePath
+  IN EFI_DEVICE_PATH_PROTOCOL  *DevicePath  OPTIONAL
```
Comment (mikebeaton) [inline Library/OcDebugLogLib/DebugPrint.c:23]: @vit9696 I am proposing making this argument OPTIONAL in this one and the new one, that seemed like the cleanest change. Is that okay?

File: Library/OcDebugLogLib/DebugPrint.c (hunk 5 above)
Comment (mikebeaton) [reply Library/OcDebugLogLib/DebugPrint.c]: I've updated and added new question on revised commit.

File: Library/OcDebugLogLib/DebugPrint.c (hunk 8 above)
Comment (vit9696) [reply Library/OcDebugLogLib/DebugPrint.c:23]: Yes.

**Comment by mikebeaton** [comment]
@vit9696 - Thank you for all the feedback. Give me a short while (might not be till tomorrow) to double-check that everything's behaving as I expect on the test systems I have here, and then I will build the docs and push to master.

**Pull request merged**